proposals/0000-allow-constraint-type-synonyms-in-deriving-heads.md

```diff
@@ -0,0 +1,202 @@
+
+# Allow constraint tuples and synonyms in typeclass instances
+
+## Motivation
+
+Most of Haskell projects do a lot of `stock` and `anyclass` deriving
+for repeating classes. Most of time there is some default list
+of classes to be defined.
+
+Making type synonym for that seems like the obvious thing to do,
+yet it does not work.
+
+There are two constructions forbidden now, but required for that usecase:
+
+* Constraint synonyms
+* Constraint tuples
+
+Synonyms worked at some moment,
+but was disabled due to wrong behaviour in some cases.
+Reasoning for that is covered in Note "Instances and constraint synonyms"
+and [issue 13267](https://gitlab.haskell.org/ghc/ghc/-/issues/13267).
+
+Main usecase only affect deriving instances head,
+but another usecase is usage of tuples/synonyms in context of any instance.
+It seems to be less common, but natural to cover in same proposal
+to allow tuples/synonyms anywhere in instance declaration,
+leading to more consistent semantics.
+
+## Proposed Change Specification
+
+Proposal requires two purely syntaxic rewrites for instance declarations
+to happen:
+
+1. Apply all type synonyms in instance declarations.
+   This should be done recursively, until no type synonym is present anymore.
+2. For any constraint tuple in derived instance head,
+   split such declarations into multiple.
+
+After that rewrites happen, existing GHC semantics is applied to result,
+with same error messages or produced code.
+
+There are two cases not covered by this rewrites:
+
+1. Using tuples in instance context leads to nested tuples.
+   But they already have the same semantics as flatten tuples,
+   so this is concern only affecting AST handling.
+2. Syntax and need for user-defined instances of class tuples is not clear.
+   So tuples in user-defined instances head remain forbidden.
+
+## Examples
+
+Main usecase:
+
+```haskell
+type ToFromJSON x = (ToJSON x, FromJSON x)
+
+-- Allowed by this proposal
+derive instance ... => (ToFromJSON MyData)
+
+-- Allowed by this proposal and working same as example above
+data MyData = MkMyData deriving (ToFromJSON)
+
+-- Should work the same with syntax allowed before proposal
+derive instance ... => ToJSON MyData
+derive instance ... => FromJSON MyData
+```
+
+Non-standalone instances do not have explicit params,
+so rewriting implementation may be little different.
+But they have just the same samantics as standalone instances with holes,
+covered by examples before, so could be handled the same.
+
+Instance context case is simpler:
+
+```haskell
+class ToFromJSON x => MyProtocolEncoding x where
+   ...
+
+data MyContainer x = ...
+
+-- Allowed by this proposal
+instance
+   (MyProtocolEncoding x, ToFromJSON (MyContainer x)) =>
+   MyContainer x where
+   ...
+
+-- Should work the same with syntax allowed before proposal
+instance
+   (MyProtocolEncoding x, ToJSON (MyContainer x), FromJSON (MyContainer x)) =>
+   MyContainer x where
+   ...
```

Comment:
`MyContainer x` is a type, not an instance, what's going on here? Did you perhaps mean to write `MyProtocolEncoding (MyContainer x)`?

Reply:
This example was replaced with simpler one. Thanks for your finding.